core/primitives/src/network.rs

```diff
@@ -9,13 +10,29 @@ use near_crypto::{KeyType, PublicKey, SecretKey, Signature};
 use crate::hash::{hash, CryptoHash};
 use crate::types::{AccountId, EpochId};
 
+/// Peer id is the public key.
+#[derive(
+    BorshSerialize, BorshDeserialize, Clone, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize,
+)]
+pub struct PeerId(Arc<PeerIdInner>);
```

Comment:
@matklad - PTAL - I'm surprised that the performance difference is so large.

Reply:
This feeld roughly right to me, tbh. The sizeof goes from 65 to just 8. Given that we use, eg, `HashMap<PeerId, _>` or `Vec<PeerId>`, the savings here seem like they can be substantial. 

What is surprising though, is that we see perf change for `calculate_distance`. That code doesn't operate with `PeerId`s at all, it's all just indices and bitmasks. We only need PeerId when we convert the result of the algorithm back to the answer. That is, the runtime of the algorithm itelf is englible in comparison to operations with PeerId to return the answer of the right shape.

This makes me think that maybe we should reconsider the repr here, and try to use a rustic impl with indices? Basically, we'd use `PeerId(u32)` internally everywhere, and use the full PublicKey repr only when sending a message? That means lifting `p2id` and `id2p` hasmaps somwhere to PeerManager (?) layer.

Reply:
> This feeld roughly right to me, tbh. The sizeof goes from 65 to just 8. Given that we use, eg, `HashMap<PeerId, _>` or `Vec<PeerId>`, the savings here seem like they can be substantial.
> 
> What is surprising though, is that we see perf change for `calculate_distance`. That code doesn't operate with `PeerId`s at all, it's all just indices and bitmasks. We only need PeerId when we convert the result of the algorithm back to the answer. That is, the runtime of the algorithm itelf is englible in comparison to operations with PeerId to return the answer of the right shape.

@matklad `Calculate_distance` calls ` self.compute_result(&mut routes, &distance)` to compute the result.
We spend most of the time converting the result to  `HashMap<PeerId, Vec<PeerId>>`

> This makes me think that maybe we should reconsider the repr here, and try to use a rustic impl with indices? Basically, we'd use `PeerId(u32)` internally everywhere, and use the full PublicKey repr only when sending a message? That means lifting `p2id` and `id2p` hasmaps somwhere to PeerManager (?) layer.

@matklad The best solution would be to keep a global structure representing list of all created `PublicKeys`. We would reduce data usage a bit, but I need to think more about it. It would be best to do this while decoding data from the network to completely avoid doing the memory allocation.

Honestly, using Arc<PublicKey> is not that but, we could save a lot of memory, by simply removing the duplicates. Though having `PeerId(u32)` , would have it easier to compute hash functions. However, that would require us to keep track when PeerId is no longer used, or risk leaking memory.

The easier way to do this, would be to override `PeerId` -> `BorshSerialize` to do a call global structure, which keep track of all duplicates. However, that has it's own issues. We probably don't want static structure for unit tests.

Reply:
Btw. `p2id` and `id2p` are not inside `RoutingTableActor`. 
Anyway, let's do other changes first, then we can talk about maybe replacing it. 

Keep in mind that using ` PeerId(u32)` by itself will not be enough. At some point you need to serialize it and convert it to `PeerId(PublicKey)`.`Will RoutingTableActor do it?` Or maybe should that be done by `PeerActor`. Though that requires inner process synchronization, which may not be a bad idea.

Other way would be to have a global cache of `PeerId<Arc<...>>`, we would use it while decoding to remove duplicates.

We have a few design choices. First, I want to do changes, that are good, and obvious how to do it.